Mimic python_client readme displaying JavaDoc examples

checketts opened the pull request:
Addresses #133

checketts commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+gauge.setToCurrentTime(); //Set to current unixtime
+
+# Increment when entered, decrement when exited. (Not implemented yet)
+inprogressRequests.trackInProgress(() -> doWork());
```
I noted this function exists exists in Python. Should I add a parallel in the Java Client?

brian-brazil replied:
I'd like to wait and see how the lambdas go before we start adding lambda-only functionality. The base variant of this doesn't make sense in Java 6.

checketts replied:
This wouldn't be lambda only, it would be Java 6 compatible, just using `Runnables` as in:

```
inprogressRequests.trackInProgress(new Runnabe() {
  run() {
    //do work
  }
}
```

I'll spin that idea into another pull request so it evolves separately and doesn't hold up the readme update.

checketts commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+
+//or (Not implemented yet)
+queueDepth.setFunction(() -> executorService.getQueue().size());
+queueDepth.setFunction(() -> executorService.getQueue().size(), "label");
```
I noted this function exists exists in Python. Should I add a parallel in the Java Client?

brian-brazil replied:
setChild is the Java equivalent.

checketts replied:
I added the setChild example, though I still think a first class method like Python exposes is deserved. I'll remove the `setFunction` callout.

brian-brazil replied:
I'd probably not mention setChild so prominently, as it may be used in appropriately. A brief mention that it exists for advanced use cases and a link to the javadoc should be enough.

brian-brazil commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
 
 [![Build Status](https://travis-ci.org/prometheus/client_java.png?branch=master)](https://travis-ci.org/prometheus/client_java)
+
+## Instrumenting
```
This should start a bit higher up I think, code examples are the most interesting thing to new users I expect.

brian-brazil commented on README.md:
````diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+There are utilities for common use cases:
+
+```java
+gauge.setToCurrentTime(); //Set to current unixtime
````
Space after //, period at end of comment

brian-brazil commented on README.md:
````diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+inprogressRequests.trackInProgress(() -> doWork());
+```
+
+A Gauge can also take its value from a callback:
````
I'd just put a brief mention of the existence of this function here, as those used to DropWizard may use it where they should be using normal direct instrumentation.

checketts replied:
Are you saying we should steer people away from this technique? I think I'm still missing what you mean by 'direct instrumentation' and how that is preferable to the DropWizard approach (I'm one of those used to DropWizard).

brian-brazil replied:
Yes, this was originally added to make a certain Spark use case easier.

The problem with the DW approach is that if often leaves things like concurrency handling up to the user, where they should really just use a Gauge (or DW Counter, which is a Prometheus Gauge).

brian-brazil commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+    .name("requests_latency_seconds").help("Request latency in seconds.").register();
+  
+  void processRequest(Request req) {
+    requestLatency.timer(() -> {
```
Mention this is java 8 only

brian-brazil commented on README.md:
````diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+}
+```
+
+There are utilities for timing code and support quantiles:
````
You should link here the discussion on quantiles, as they're not aggregatable and relatively slow.

checketts replied:
Do you mean this one? https://prometheus.io/docs/practices/histograms/#quantiles

brian-brazil replied:
Yes

brian-brazil commented on README.md:
````diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+```
+
+The default buckets are intended to cover a typical web/rpc request from milliseconds to seconds.
+They can be overridden with the `buckets()` method on the `Histogram.Builder`.
````
Should we link down to the javadoc on these?

brian-brazil commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+     .labelNames("method").register();
+  
+  void processGetRequest() {
+    requests.labels("get").inc();
```
Because the label value here is static, what you really should be doing is putting the labels() result in a class variable and then calling inc on it. This is so it gets initialised, and it's also faster.

checketts replied:
This example just mimics the JavaDoc. I would like to create a separate section about registering and efficient usage on a followup PR that discusses why to use `static final` and registering metrics and what `Child` is. Is it OK if we leave this as is?

brian-brazil replied:
Ok

brian-brazil commented on README.md:
````diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+}
+```
+
+### Included Exporters
````
Collectors

brian-brazil replied:
There's the logging ones too

checketts replied:
OK Added a logback example.

brian-brazil commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+
+### Included Exporters
+
+The Java client includes exporters for Garbage Collection, Memory Pools, JMX, and Thread counts.
```
Just point to DefaultExports

brian-brazil commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+
+### HTTP
+
+Metrics are usually exposed over HTTP, to be read by the Prometheus server.
```
I'd include a minimal servlet example, and link off to the docstrings for the rest.

checketts replied:
I'm not sure what the minimal servlet example would look like in a readme. It would just be a `web.xml` mapping a servlet. Even the DropWizard equivalent just says '... you can map the servlet to whatever path you see fit.' (http://metrics.dropwizard.io/3.1.0/getting-started/#reporting-via-http)

brian-brazil replied:
Borrow from https://www.robustperception.io/instrumenting-java-with-prometheus/

brian-brazil commented on README.md:
```diff
@@ -47,3 +47,297 @@ Details for those wishing to develop the library can be found on the [wiki](http
+Sometimes it is not possible to directly instrument code, as it is not
+in your control. This requires you to proxy metrics from other systems.
+
+To do so you need to create a custom collector (which will require to be registered as a normal metric), for example:
```
will need to be

checketts commented:
I think I've addressed all the comments. I want to do a follow up PR with the following topics:

- Efficiently creating/registering a metrics (`static final` class members)
- What 'Child' means in regards to a metric
- Summarization and links to the metrics naming and recommendations around tracking success failure in a Counter and not a Histogram/Summary

checketts commented on README.md:
```diff
@@ -47,3 +47,324 @@ Details for those wishing to develop the library can be found on the [wiki](http
+
+## Instrumenting
+
+Four types of metric are offered: Counter, Gauge, Summary and Histogram.
```
Thanks for the proofread!

brian-brazil commented on README.md:
````diff
@@ -34,16 +36,342 @@ version can be found on in the maven repository for
+It also serves as a simple example of how to write a custom endpoint.
+
+```java
+context.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
````
Include the few lines before this, enough to get it running.

checketts replied:
You mean this?

> Setup
>
> There’s a small bit of setup work that need to be done once, no matter how many metrics you have. To expose the metrics used in your code, we add the Prometheus servlet to our Jetty server:

brian-brazil replied:
I mean the few lines of code you need with that to create the server and context.

brian-brazil commented on README.md:
````diff
@@ -34,16 +36,344 @@ version can be found on in the maven repository for
+gauge.setToCurrentTime(); // Set to current unixtime.
+```
+
+As an advanced use case, a `Gauge` can also take its value from a callback. Keep in mind that the default `inc()`, `dec()` and `set()`
````
As already discussed, link to the javadoc here rather than providing an example.

checketts replied:
Are you requesting I remove the `setChild` example?  There is no example on Gauge in its JavaDocs and no one will be hunting in parent classes for the example. I know I wouldn't have found it without your help in #178 . 

I'll remove the example, but I still can't wrap my head around why you think it shouldn't have up front documentation (aside from the python equivalent `setFunction` being a bit better named than `setChild`) and reminding people that they need to account for concurrency.

brian-brazil replied:
> There is no example on Gauge in its JavaDocs and no one will be hunting in parent classes for the example.

Link to the example we have.

> I still can't wrap my head around why you think it shouldn't have up front documentation

Basically dropwizard has taught some developers that this is the way to handle gauges, whereas with Prometheus client libraries this is rarely the appropriate way to do things. Thus I want to make it less likely that users will tie themselves up in knots trying to use this dropwizard-style.

checketts commented:
I believe all points have been addressed.

brian-brazil commented:
Great! Thanks for working through this, it'll make life easier for a lot of people.

oliver006 commented:
🎉
